Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,128 @@
+to point-free topology and order theory. Applications are found throughout logic,
+quantum mechanics, and computer science. Intuitively, as described by [vickers1989],
+open sets of a topology form a frame, and can be considered as modeling what is `observable`
+in a system. Quantales come into play when making an observation may change the system itself.
+Traditionally, one would write `x ⊓ y` to describe making observations `x` and `y` at the same
+time, but when the order of observation becomes important due to changes in the system caused
+by the observation, it makes more sense to write `x * y` or `x + y` and consider `*` or `+` as
+a possibly non-commutative monoid.
+
+## Main definitions
+
+* class `Quantale`: a monoid distributing over a complete sup-semilattice.
+  I.e satisfying `x * (sSup s) = ⨆ y ∈ s, x * y` and `(sSup s) * y = ⨆ x ∈ s, x * y`
+
+Conceptually, every complete sup-semilattice is of course a complete lattice.
+However, when considering morphisms between quantales, usually only the supremum
+is required to be preserved. Hence, it is conseptually cleaner to define it as a
+complete sup-semilattice. (In literature, both variants occur.)
+
+* `is_integral`: a quantale for which `1 = ⊤`
+* `is_comm`: a commutative quantale, satisfying `x * y = y * x`
+* `is_idem`: an idempotent quantale, satisfying `x * x = x`
+
+## Naming conventions
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the left-residuation of `y` over `x`;
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the right-residuation of `y` over `x`;
+
+## References
+
+[Topology via Logic][vickers1989]
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale> (Categorical definition - not followed here)
+
+## TODO
+
+Additive quantales and multiplicative quantales both occur in literature.
+For sequences of actions, usually a multiplicative quantale is used, but for
+describing timing of actions (for example) the additive quantale of max-plus algebra
+a.k.a. tropical algebra, can be used. Therefore, extending the definitions with
+addition is - I think - relevant future work.
+
+-/
+
+/-- A quantale is a monoid distributing over a complete sup-semilattice.
+
+    Technically, I would prefer to define it as
+    `class Quantale (α : Type*) extends Monoid α, CompleteSemilatticeSup α where ...`
+    however, if I do that currently, the elements ⊤ and ⊥ are not automatically defined.
+    I could do that separately, but it seems more sensible to push the definition
+    of those to CompleteSemmilatticeSup in the library, as they are already making
+    sense there and having them should not break anything (in theory).
+
+    The reason not to consider a quantale as a complete lattice, is because the
+    morphisms between quantales usually only preserve the join and not the inf.
+    In fact, when seen as a generalization of locales, the monoid multiplication
+    usually takes the place of the inf.
```

Comment:
I think you *really* don't want to go the `CompleteSemilatticeSup` route. Indeed, you will lose out on *many* lemmas about `CompleteLattice`s that way, which will just cause pain. Even if in the theory of quantales one never needs the `inf`, it's still useful to have the full structure around. If a use wants to only provide the `CompleteSemilatticeSup` structure and then have it upgraded to a `CompleteLattice` for free, we have a constructor that does that (which will make the defeqs for `inf` and `sInf` terrible, but as you indicate, it's possible that won't matter).

As for the morphisms, there's no problem: when you get around to defining them, just define them to only preserve the join.

Reply:
Question is why those lemmas that we would loose have not been formulated on CompleteSemiLattices in the first place. If they really do not hold for CompleteSemiLattices, then it is doubtful whether you would want to use them, since you are doing something conceptually strange (categorically speaking).

Reply:
I see. So in fact you only have the CompleteSemilatticeSup in order to be able to prove that it is a CompleteLattice, and then build everything on that.

I first thought it was conceptually nicer to `only have a CompleteSemilatticeSup' since my own intuition for using Quantales is that the `inf' is not meaningful. But considering it meaningful or not is, I now realize, a bias on my part - based on how I happen to think of them at this point in time. If I let go of that, and in a library you must be able to support all kinds of uses, then it makes sense to just add everything you can reasonably define, and you end up with complete lattices. Thanks... I learned something today :-)